hazelcast/src/jepsen/hazelcast.clj

```diff
@@ -343,22 +595,209 @@
          :read (assoc op :type :ok, :value (into (sorted-set) (.get m "hi")))))
 
      (teardown! [this test]
-       (.terminate conn)))))
+       (.shutdown conn)))))
 
 (defn map-workload
   "A workload for map tests, with the given client options."
   [client-opts]
-  {:client    (map-client client-opts)
-   :generator (->> (range)
-                   (map (fn [x] {:type  :invoke
-                                 :f     :add
-                                 :value x}))
-                   gen/seq
-                   (gen/stagger 1/10))
+  {:client          (map-client client-opts)
+   :generator       (->> (range)
+                         (map (fn [x] {:type  :invoke
+                                       :f     :add
+                                       :value x}))
+                         gen/seq
+                         (gen/stagger 1/10))
    :final-generator (->> {:type :invoke, :f :read}
                          gen/once
                          gen/each)
-   :checker   (checker/set)})
+   :checker         (checker/set)})
+
+
+
+(defn parseLine [line]
+  (let
+    [tokens (.split line CLIENT_ID_SEPARATOR)]
+    [(:value (clojure.edn/read-string (nth tokens 2))) (.trim (nth tokens 1))]))
+
+(def clientUidsToClientNames (memoize (fn []
+                            (apply array-map
+                                   (flatten
+                                     (map parseLine
+                                          (.split (:out (sh "grep" CLIENT_ID_SEPARATOR "store/latest/jepsen.log")) "\n")
+                                          ))))))
+
+(defn getClient [op]
+  (let [val (:value op)]
+    (if (map? val) (:client val) (get (clientUidsToClientNames) (:value op)))))
+
+(defrecord ReentrantMutex [owner lockCount]
+  Model
+  (step [this op]
+    (let [client (getClient op)]
+      (if (nil? client)
+        (do
+          (info "no owner!")
+          (knossos.model/inconsistent "no owner!")
+          )
+        (condp = (:f op)
+          :acquire (if (and (< lockCount REENTRANT_LOCK_ACQUIRE_COUNT) (or (nil? owner) (= owner client)))
+                     (ReentrantMutex. client (+ lockCount 1))
+                     (knossos.model/inconsistent (str "client: " client " cannot " op " on " this))
+                     )
+          :release (if (or (nil? owner) (not= owner client))
+                     (knossos.model/inconsistent (str "client: " client " cannot " op " on " this))
+                     (ReentrantMutex. (if (= lockCount 1) nil owner) (- lockCount 1))
+                     )
+          )
+        )
+      )
+    )
+
+  Object
+  (toString [this] (str "owner: " owner ", lockCount: " lockCount)))
+
+(defn createReentrantMutex []
+  "A single reentrant mutex responding to :acquire and :release messages"
+  (ReentrantMutex. nil 0))
+
+
+
+(defrecord OwnerAwareMutex [owner]
+  Model
+  (step [this op]
+    (let [client (getClient op)]
```

Comment:
Ditto here; I think you can get rid of external client IDs and use processes IDs instead.